sympy/polys/polyfuncs.py

```diff
@@ -181,28 +196,128 @@ def horner(f, *gens, **args):
+        if 'wrt' in args:
+            F, gens = Poly(F, gen), gens[1:]
+
+            for coeff in F.all_coeffs():
+                form = form*gen + horner(coeff, *gens, **args)
+        elif method == 'random': # randomly select a variable to factor
+            N = len(F.gens)**2
+            minmul = len(str(F.as_expr))
+            for i in range(N):
+                gen = choice(F.gens)
+                F2 = Poly(F, gen)
+                form = S.Zero
+                args['method'] = 'random-helper'
+                gens = tuple(d for d in gens if d != gen)
+                for coeff in F2.all_coeffs():
+                    form = form*gen + horner(coeff, *gens, **args)
+                nmul = nmult(form)
+                if nmul < minmul:
+                    minform = form
+                    minmul = nmul
+            form = minform
+        elif method == 'random-helper':
+            gen = choice(F.gens)
+            F = Poly(F, gen)
+            form = S.Zero
+            gens = tuple(d for d in gens if d != gen)
+            for coeff in F.all_coeffs():
+                form = form*gen + horner(coeff, *gens, **args)
+        elif method == 'random-greedy': # a mixture of random and greedy
+            N = len(F.gens)**2
+            s = str(F.as_expr)
+            ls = len(s)//2
+            minmul = len(s)
+            for i in range(N):
+                maxgen = max(((s.count(str(gen)),gen) for gen in F.gens),key=lambda x:x[0])[1]
+                gen = choice(list(F.gens)+[maxgen]*ls)
+                F2 = Poly(F, gen)
+                form = S.Zero
+                args['method'] = 'random-greedy-helper'
+                gens = tuple(d for d in gens if d != gen)
+                for coeff in F2.all_coeffs():
+                    form = form*gen + horner(coeff, *gens, **args)
+                nmul = nmult(form)
+                if nmul < minmul:
+                    minform = form
+                    minmul = nmul
+            form = minform
+        elif method == 'random-greedy-helper':
+            s = str(F.as_expr)
+            ls = len(s)//2
+            maxgen = max(((s.count(str(gen)),gen) for gen in F.gens), key=lambda x:x[0])[1]
+            gen = choice(list(F.gens)+[maxgen]*ls)
+            F = Poly(F, gen)
+            form = S.Zero
+            gens = tuple(d for d in gens if d != gen)
+            for coeff in F.all_coeffs():
+                form = form*gen + horner(coeff, *gens, **args)
+        elif method == 'greedy': # factor out the variable that occurs most often
+            s = str(F.as_expr)
+            gen = max(((s.count(str(gen)),gen) for gen in F.gens), key=lambda x:x[0])[1]
+            F = Poly(F, gen)
+            form = S.Zero
+            gens = tuple(d for d in gens if d != gen)
+            for coeff in F.all_coeffs():
+                form = form*gen + horner(coeff, *gens, **args)
+        elif method == 'best': # exhaustive search
+            minmul = len(str(F.as_expr))
```

Comment:
Again, don't use strings. There is `count_ops` to count the number of operations.